File: src/domain/models/Transaction.ts
import Decimal from 'decimal.js';

export enum TransactionType {
  BUY = 'BUY',
  SELL = 'SELL',
  DIVIDEND = 'DIVIDEND',
  SPLIT = 'SPLIT',
  FEE = 'FEE',
  TRANSFER = 'TRANSFER',
  INTEREST = 'INTEREST',
}

export enum Broker {
  ROBINHOOD = 'ROBINHOOD',
  SCHWAB = 'SCHWAB',
  CHARLES_SCHWAB = 'CHARLES_SCHWAB',
  GENERIC = 'GENERIC',
  MANUAL = 'MANUAL',
}

export interface Transaction {
  id: string; // UUID
  date: Date;
  symbol: string;
  type: TransactionType;
  quantity: Decimal;
  price: Decimal;
  fees: Decimal;
  currency: string;
  broker?: string;
  rawData?: string;
  notes?: string;
  tags?: string[];
  rating?: number;
}

export interface RawTransaction {
  symbol: string;
  transaction_type: TransactionType;
  transaction_date: string;
  settle_date?: string;
  quantity: string;
  price: string;
  fees: string;
  broker: Broker;
  raw_data: string;
  notes?: string;
}

export function createTransaction(
  raw: RawTransaction,
  _batchId?: string
): Transaction {
  return {
    id: crypto.randomUUID(),
    date: new Date(raw.transaction_date),
    symbol: raw.symbol,
    type: raw.transaction_type,
    quantity: new Decimal(raw.quantity),
    price: new Decimal(raw.price),
    fees: new Decimal(raw.fees),
    currency: 'USD',
    broker: raw.broker,
    rawData: raw.raw_data,
    notes: raw.notes,
  };
}

File: backend/app/models/transaction.py
from __future__ import annotations
from pydantic import BaseModel, Field
from typing import Any
from enum import Enum
from datetime import date, datetime
import uuid

class TransactionType(str, Enum):
    BUY = "BUY"
    SELL = "SELL"
    DIVIDEND = "DIVIDEND"
    INTEREST = "INTEREST"
    TRANSFER = "TRANSFER"
    FEE = "FEE"
    # Add other types as needed

class ImportBatchBase(BaseModel):
    broker: str
    filename: str
    imported_at: datetime
    row_count: int

class ImportBatch(ImportBatchBase):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))

class TransactionBase(BaseModel):
    import_batch_id: str
    symbol: str
    transaction_type: str
    transaction_date: date
    settle_date: date | None = None
    quantity: float
    price: float
    fees: float
    total_amount: float
    broker: str
    account_id: str | None = None
    notes: str | None = None
    raw_data: str # Storing JSON string or CSV row content
    content_hash: str

class Transaction(TransactionBase):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    created_at: datetime = Field(default_factory=datetime.now)
